core/libp2p/security/plaintext/exchange_message_marshaller.hpp

```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_EXCHANGE_MESSAGE_MARSHALLER_HPP
+#define KAGOME_EXCHANGE_MESSAGE_MARSHALLER_HPP
+
+#include <vector>
+
+#include "libp2p/security/plaintext/exchange_message.hpp"
+
+namespace libp2p::security::plaintext {
+
+  /**
+   * Performs serializing of a Plaintext exchange message to Protobuf format and
+   * deserializes it back
+   */
+  class ExchangeMessageMarshaller {
+   public:
+    virtual ~ExchangeMessageMarshaller() = default;
+
+    /**
+     * @param msg exchange message to be marshalled to Protobuf
+     * @returns a byte array containing a Protobuf representation of an exchange
+     * message
+     */
+    virtual outcome::result<std::vector<uint8_t>> marshal(
+        const ExchangeMessage &msg) const = 0;
+
+    /**
+     * @param msg_bytes a byte array containing a Protobuf representation of an
+     * exchange message
+     * @returns a deserialized exchange message
+     */
+    virtual outcome::result<ExchangeMessage> unmarshal(
+        gsl::span<const uint8_t> msg_bytes) const = 0;
+  };
```

Comment:
I don't like this. We need better structure for marshalling/unmarshalling...
Specifically, I don't like that for **each message type** we need to create an interface, its implementation... Ugh

Any suggestions?

Reply:
At least write some utility functions for common operations like convenient conversion from/to vector. To do more than that... I think with our current clean architecture principle it would be hard to avoid interface/implementation thing, as some of them depend on others, but at least if most common operations would be extracted to methods an implementation for a message could be kept quite short, I suppose.

Reply:
I suggest we might use scale :)

Reply:
I like that idea.

Reply:
It was sort of joke. I think we should use same method as in other implementations to stay compatible. If it is scale, great - less efforts, more clear.

Reply:
Sure.

Reply:
Maybe we can do the same "trick" as with Scale? Define two template operators (or methods), which will be resolved into the corresponding implementations for each type we want to marshal?